### PR TITLE
[Merged by Bors] - feat: port Algebra.Module.Submodule.Basic

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by jcommelin** [comment]
Please rebase on `master`.

**Comment by Vierkantor** [comment]
I'm going in 10 to 15 minutes. Everything should build and lint correctly, we need to run the other fixing scripts and check the names and docstrings.

File: Mathlib/Algebra/Module/Submodule/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+  rfl
+#align submodule.coe_subtype Submodule.coeSubtype
+
+theorem injective_subtype : Injective p.subtype :=
+  Subtype.coe_injective
+#align submodule.injective_subtype Submodule.injective_subtype
+
+/-- Note the `add_submonoid` version of this lemma is called `AddSubmonoid.coe_finset_sum`. -/
+-- porting note: removing the `@[simp]` attribute since it's lterally `AddSubmonoid.coe_finset_sum`
+@[simp]
```
Comment (Vierkantor) [inline Mathlib/Algebra/Module/Submodule/Basic.lean:382]:
```suggestion
```

File: Mathlib/Algebra/Module/Submodule/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+
+/-- Even though `p.restrictScalars S` has type `Submodule S M`, it is still an `R`-module. -/
+instance restrictScalars.origModule (p : Submodule R M) : Module R (p.restrictScalars S) :=
+  (by infer_instance : Module R p)
+#align submodule.restrict_scalars.orig_module Submodule.restrictScalars.origModule
+
+instance (p : Submodule R M) : IsScalarTower S R (p.restrictScalars S)
+    where smul_assoc r s x := Subtype.ext <| smul_assoc r s (x : M)
+
+/-- `restrict_scalars S` is an embedding of the lattice of `R`-submodules into
```
Comment (jcommelin) [inline Mathlib/Algebra/Module/Submodule/Basic.lean:437]:
```suggestion
/-- `restrictScalars S` is an embedding of the lattice of `R`-submodules into
```

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**